### PR TITLE
GUVNOR-3102: Prevent metadata key from using special characters

### PR DESCRIPTION
**Pull request by jomarko**
@karreiro please take a look when you have a moment.

### CODE REVIEW

**Comment by karreiro** [comment]
+1 (when green) Thanks @jomarko

**Comment by karreiro** [comment]
Jenkins, please restest this.

**Comment by karreiro** [comment]
Jenkins, please retest this.

**Comment by jomarko** [comment]
jenkins retest this please

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/wizard/column/pages/MetaDataColumnPage.java [hunk 1]
```diff
@@ -19,28 +19,29 @@
...
     private View view;
+    private RegExp containWhiteSpacePattern;
 
     @Inject
     public MetaDataColumnPage(final View view,
                               final TranslationService translationService) {
         super(translationService);
 
         this.view = view;
+        this.containWhiteSpacePattern = RegExp.compile("\\s");
```
Comment (jomarko) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/wizard/column/pages/MetaDataColumnPage.java]: Very good point. You are definitely right @jhrcek there are other prohibited characters for metadata like '*' or '#'. However I am not sure what strings are accepted by the compiler. @karreiro @manstis any ideas what set of characters should be prohibited for the metadata?

Comment (manstis) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/wizard/column/pages/MetaDataColumnPage.java]: @jomarko TBH I didn't know there was _any_ restriction: https://docs.jboss.org/drools/release/6.5.0.Final/drools-docs/html/ch08.html#d0e8397 makes no mention of reserved characters. @mariofusco can you advise the restrictions on rule meta-data?

Comment (mariofusco) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/wizard/column/pages/MetaDataColumnPage.java]: @manstis @jomarko @jhrcek Are you speaking about the key or the value of the metadata? In the first case it has to be a valid Java identifier while in the second there are no restrictions (I just checked that also the space works).

Comment (jomarko) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/wizard/column/pages/MetaDataColumnPage.java]: @mariofusco we were discussing the key. Thanks.

Comment (jomarko) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/wizard/column/pages/MetaDataColumnPage.java]: @manstis @karreiro I have some feeling that I saw already somewhere util method that check if string is valid java variable? Any ideas where to findt it?

Comment (manstis) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/wizard/column/pages/MetaDataColumnPage.java]: @jomarko You can use ```[ValidationService.evaluateJavaIdentifiers(...)](https://github.com/kiegroup/kie-wb-common/blob/master/kie-wb-common-services/kie-wb-common-services-backend/src/main/java/org/kie/workbench/common/services/backend/validation/ValidationServiceImpl.java#L114)```

**Comment by jomarko** [comment]
@manstis please, could you please help me? The dependency injection fails, I tried to add dependency for `kie-wb-common-services-backend` into `drools-wb-guided-dtable-editor-client` module, however then fails maven rule enforcer.

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/wizard/column/pages/MetaDataColumnPage.java [hunk 2]
```diff
@@ -35,12 +34,16 @@
 
     private View view;
 
+    private ValidationService validationService;
```
Comment (manstis) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/wizard/column/pages/MetaDataColumnPage.java]: @jomarko You'll need to inject ```Caller<ValidationService> service```

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/wizard/column/pages/MetaDataColumnPage.java [hunk 3]
```diff
@@ -74,18 +95,16 @@ public void setMetadata(String metadata) {
...
 
         void showError(String errorMessage);
 
+        void hideError();
+
         void clear();
     }
+
+    private boolean isValidIdentifier(String identifier) {
+        return validationService.evaluateJavaIdentifiers(new String[] {identifier}).get(identifier);
```
Comment (manstis) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/wizard/column/pages/MetaDataColumnPage.java]: You'll also need to change this to do something like ```service.call( () -> {...}).evaluateJavaIdentifiers(..)```

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/wizard/column/pages/MetaDataColumnPage.java (hunk 2 above)
Comment (jomarko) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/wizard/column/pages/MetaDataColumnPage.java]: Great, thanks.

**Comment by jomarko** [comment]
screenshot from 2017-05-26 13-39-47
screenshot from 2017-05-26 13-39-25
screenshot from 2017-05-26 13-39-08
screenshot from 2017-05-26 13-39-00

I hope this is finally done.

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/wizard/column/pages/MetaDataColumnPageTest.java [hunk 4]
```diff
@@ -70,27 +86,54 @@ public static void setupPreferences() {
...
+    @Before
+    public void setUp() throws Exception {
+        when(translationService.format(GuidedDecisionTableErraiConstants.MetaDataColumnPage_MetadataNameEmpty))
+                .thenReturn("empty");
+        when(translationService.format(GuidedDecisionTableErraiConstants.MetaDataColumnPage_ThatColumnNameIsAlreadyInUsePleasePickAnother))
+                .thenReturn("alreadyUsed");
+        when(translationService.format(GuidedDecisionTableErraiConstants.MetaDataColumnPage_IsNotValidIdentifier))
+                .thenReturn("isNotValid");
+
+        when(validationServiceCaller.call(any())).thenReturn(validationService);
```
Comment (manstis) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/wizard/column/pages/MetaDataColumnPageTest.java]: Using ```CallerMock``` (in ```uberfire-testing-utils```) is better/easier. You'd do ```validationServiceCaller = new CallerMock(validationService);``` and then your test won't need the ```RemoteCallback``` captor.

Comment (jomarko) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/wizard/column/pages/MetaDataColumnPageTest.java]: @manstis great hint. I always forget for this `CallerMock`. Thanks.

**Comment by jomarko** [comment]
@manstis done. I wait for other suggestions, comments, anything.

**Comment by manstis** [comment]
@jhrcek OK with you?

**Pull request merged**